Added changes pertaining to #69 and #74

denislavh opened the pull request:
Added SettingsBase and DateTimeSettings for closing #69
Added ISettingsService and SettingsService for closing #74

denislavh committed:
Added SettingsBase and DateTimeSettings for closing #69
Added ISettingsService and SettingsService for closing #74

chrisK00 commented on subtrack.DAL/Entities/DateTimeSetting.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class DateTimeSetting : SettingsBase
+    {
+        public const string Key = "LastAutoPaymentTimeStamp";
```
My bad, but we should probably name this LastAutoPaymentTimeStampKey since DateTimeSetting can be used for any setting

chrisK00 commented on subtrack.DAL/SubtrackDbContext.cs:
```diff
@@ -6,8 +6,13 @@ namespace subtrack.DAL;
     }
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        modelBuilder.Entity<SettingsBase>().HasDiscriminator<string>("settings_type").HasValue<DateTimeSetting>(nameof(DateTimeSetting));
```
My preference: place .HasValue on a new line, alternatively every "." on a new line

chrisK00 commented on subtrack.MAUI/Services/SettingsService.cs:
```diff
@@ -0,0 +1,29 @@
+        private readonly SubtrackDbContext _context;
+        public async Task<T> GetByIdAsync<T>(string id) where T : SettingsBase
+        {
+            return (T)await _context.Settings.FirstOrDefaultAsync(x => x.Id == id);
```
Make use of .OfType<T> instead of casting

chrisK00 commented on subtrack.MAUI/Services/SettingsService.cs:
```diff
@@ -0,0 +1,29 @@
+        public async Task UpdateAsync<T>(T setting) where T : SettingsBase
+
+        {
+            var sett = GetByIdAsync<SettingsBase>(setting.Id);
```
We actually need to make an update here, call _context.Update. It should handle the update for you

denislavh committed:
Changed const inside DateTimeSettings.
Removed casting from SettingsService and made use of OfType.
Redone Update Method inside SettingsService.

chrisK00 commented on subtrack.MAUI/Services/SettingsService.cs:
```diff
@@ -0,0 +1,27 @@
+        public async Task UpdateAsync<T>(T setting) where T : SettingsBase
+
+        {
+            _context.Update(setting);
```
Does .Update call .SaveChanges?

denislavh replied:
My bad, from what I understand it begins tracking the specified entity in modified state but no changes are made until SaveChangesAsync() is actually called. https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.update?view=efcore-7.0 https://learn.microsoft.com/en-us/ef/core/change-tracking/#how-to-track-entities

denislavh committed:
Added .SaveChangesAsync() to SettingsService.

chrisK00 reviewed (approved):
Merge 🔥
